Add a new transpiler pass to remove all barriers

peendebak opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Add pass to remove all barriers.

### Details and comments

Implementation is based on the `MergeAdjacentBarriers` pass.
A minimal example:
```
from qiskit import QuantumCircuit
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import RemoveBarriers

circuit = QuantumCircuit(1)
circuit.barrier()
circuit.x(0)
circuit.barrier()
circuit.h(0)
pm = PassManager(RemoveBarriers())
circuit = pm.run(circuit)
circuit.draw()
```

peendebak commented:
@1ucian0 Could please you review this PR?

ajavadia reviewed (commented):
Looks good to me, can you add a release note (using `reno`)?
I think in principle having this pass can be handy, but it is not semantics preserving as barriers can have semantic information. Maybe add a note to that effect as well.

peendebak commented:
@ajavadia I updated the docstring to note the transformation does not preserve semantics. Can I describe this more precisely? Any transformation that is not the identity does not preserve semantics right? And the removal of barriers leaves the unitary intact, so those semantics are not changed.

I added release notes as well.

1ucian0 commented:
the docstring seems to have an error.

peendebak commented:
@1ucian0 The error message generated is not very helpful. It seems to be related to `plot_state_paulivec`, which is not related at all to this PR.

![image](https://user-images.githubusercontent.com/17724047/119187601-d4447f00-ba79-11eb-8917-4921d7b2d0cb.png)

I cannot reproduce the error locally. Can you give suggestions on how to tackle this?

ajavadia commented:
> Any transformation that is not the identity does not preserve semantics right?
Hmm I guess it depends on the definition of semantics. If you only consider the unitary then yes it's semantics-preserving. But I meant to say that barriers are not just sugar in the circuit; they mean something (for example they indicate some user's intent about two gates not merging). So by removing those you are removing that extra information on the circuit. 

I suppose you could argue that collapsing two back-to-back X gates into identity is also changing something about the circuit, for example the type of noise that occurs on it during execution. But the semantics of OpenQASM allow this, but does not allow barrier removal. I only accepted it here because it can be a useful utility for users, but I wanted that information included.

These types of "valid" compiler transformations that tie into the semantics of the QASM language are discussed a bit in the [OpenQASM 3 paper](https://arxiv.org/abs/2104.14722), especially in the delay and timing section.

peendebak commented:
> > Any transformation that is not the identity does not preserve semantics right?
> 
> Hmm I guess it depends on the definition of semantics. If you only consider the unitary then yes it's semantics-preserving. But I meant to say that barriers are not just sugar in the circuit; they mean something (for example they indicate some user's intent about two gates not merging). So by removing those you are removing that extra information on the circuit.
> 
> I suppose you could argue that collapsing two back-to-back X gates into identity is also changing something about the circuit, for example the type of noise that occurs on it during execution. But the semantics of OpenQASM allow this, but does not allow barrier removal. I only accepted it here because it can be a useful utility for users, but I wanted that information included.
> 
> These types of "valid" compiler transformations that tie into the semantics of the QASM language are discussed a bit in the [OpenQASM 3 paper](https://arxiv.org/abs/2104.14722), especially in the delay and timing section.

@ajavadia Thanks for clarifying. I added type annotations and fixed a typo. PR should be remove for merge now.

1ucian0 commented on docs/apidocs/terra.rst:
```diff
@@ -33,6 +33,5 @@ Qiskit Terra API Reference
    transpiler_preset
    utils
    validation
-   visualization
```
Is this change related?

eendebakpt replied:
No, I will remove it.

1ucian0 commented on qiskit/transpiler/passes/utils/remove_barriers.py:
```diff
@@ -0,0 +1,56 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
```suggestion
# (C) Copyright IBM 2017, 2021.
```

1ucian0 commented on qiskit/transpiler/passes/utils/remove_barriers.py:
```diff
@@ -0,0 +1,56 @@
+        new_dag = dag._copy_circuit_metadata()
+
+        # go over current nodes, and add them to the new dag
+        for node in dag.topological_op_nodes():
+            if node.name == "barrier":
+                pass
+            else:
+                # copy the condition over too
+                new_dag.apply_operation_back(node.op, qargs=node.qargs, cargs=node.cargs)
```
Why not `dag.remove_all_ops_named("barrier")`?

eendebakpt replied:
@1ucian0 I as not aware of this method, but that is indeed better. I updated the code. 

One question: the method `dag.remove_all_ops_named` removes the operations in-place. Is that OK? I see that the `run()` method for many passes returns a new dag, but the argument dag is modified inplace as well. In the base class for passes, the run method is not annotated with a return value:

https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/basepasses.py#L70

1ucian0 replied:
I dont think we have any warranty about input preservation. I was trying to play a bit with the pass and it seems like everything seems fine.

1ucian0 reviewed (dismissed):
I think `DAGCircuit.remove_all_ops_named()` is exactly for this kind of situations (even maybe better performance?)

1ucian0 commented on test/python/transpiler/test_remove_barriers.py:
```diff
@@ -0,0 +1,57 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
```suggestion
# (C) Copyright IBM 2017, 2021.
```

1ucian0 commented on qiskit/transpiler/passes/utils/remove_barriers.py:
```diff
@@ -0,0 +1,48 @@
+
+        .. jupyter-execute::
+
+            from qiskit import QuantumCircuit
+            from qiskit.transpiler import PassManager
+            from qiskit.transpiler.passes import RemoveBarriers
+
+            circuit = QuantumCircuit(1)
+            circuit.x(0)
+            circuit.barrier()
+            circuit.h(0)
+
+            pm = PassManager(RemoveBarriers())
+            circuit = pm.run(circuit)
```
You can call a pass directly, no need to build a pass manager.
```suggestion
            from qiskit import QuantumCircuit
            from qiskit.transpiler.passes import RemoveBarriers

            circuit = QuantumCircuit(1)
            circuit.x(0)
            circuit.barrier()
            circuit.h(0)

            circuit = RemoveBarriers()(circuit)
```

mtreinish commented on releasenotes/notes/add-remove-barriers-pass-dab69660f61e9d75.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add pass RemoveBarriers to remove all barriers in a circuit.
```
```suggestion
    Added a new transpiler pass, :class:`~qiskit.transpiler.passes.RemoveBarriers`,
    to :mod:`qiskit.transpiler.passes`. This pass is used to remove all barriers in a
    circuit.
```
